gcsmigrate/shared.py

```diff
@@ -0,0 +1,354 @@
+
+    def readinto(self, b):
+        num = self.wrap.readinto(b)
+        if num is not None and num > 0:
+            self.hasher.update(b[0:num])
+        return num
+
+
+def make_bucket_map(buckets):
+    """Construct a hash from a buckets source string of the form a:b,c:d."""
+    retval = {}
+    for bucket in buckets.split(","):
+        source, target = bucket.split(":")
+        retval[source] = target
+    return retval
+
+
+def hash_fileobj(fileobj, hasher):
+    """Efficiently hash a file object using the provided hasher."""
+    fileobj.seek(0)
+    block = fileobj.read(BUFSIZE)
+    while block:
+        hasher.update(block)
+        block = fileobj.read(BUFSIZE)
+    fileobj.seek(0)
+    return hasher
+
+
+def md5_fileobj_hex(fileobj):
+    """Efficiently calculate the MD5 sum for a file object.
+
+    Returns a hex string.
+    """
+    return hash_fileobj(fileobj, hashlib.md5()).hexdigest()
+
+
+def md5_fileobj_b64(fileobj):
+    """Efficiently calculate the MD5 sum for a file object.
+
+    Returns a Base64 encoded string.
+    """
+    return base64.b64encode(hash_fileobj(fileobj, hashlib.md5()).digest()).decode(
+        "utf-8"
+    )
+
+
+def sha1_fileobj_hex(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object.
+
+    Returns a hex string.
+    """
+    return hash_fileobj(fileobj, hashlib.sha1()).hexdigest()
+
+
+def sha1_fileobj_b64(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object.
+
+    Returns a Base64 encoded string.
+    """
+    return base64.b64encode(hash_fileobj(fileobj, hashlib.sha1()).digest()).decode(
+        "utf-8"
+    )
+
+
+class MogileFile:
+    """Represents a file stored in mogile."""
+
+    def __init__(self, fid: int, dkey: str, length: int):
+        """Initialize a MogileFile.
+
+        Arguments are as stored in the mogile `file` database table.
+        """
+        self.length = length
+        self.fid = fid
+        self.dkey = dkey
+        if dkey == "test" or dkey[36:] == ".tmp":
+            # These seem to be old junk leftover from failed ingests.
+            # Check later.
+            self.skip = True
+            self.sha1sum = None
+            self.mogile_bucket = None
+        else:
+            self.skip = False
+            self.sha1sum = dkey[0:40]
+            self.mogile_bucket = dkey[41:]
+            assert len(self.sha1sum) == 40, f"bad sha1sum length for {self.dkey}"
+            assert dkey[40] == "-", f"bad dkey for {self.dkey}"
+            assert len(self.mogile_bucket) > 1, f"bad mogile-bucket for {self.dkey}"
+
+    def __eq__(self, other):
+        """Equality check."""
+        if not isinstance(other, MogileFile):
+            # Delegate comparison to the other instance's __eq__.
+            return NotImplemented
+        return (
+            self.dkey == other.dkey
+            and self.fid == other.fid
+            and self.length == other.length
+        )
+
+    @classmethod
+    def parse_row(cls, row: dict):
```

Comment:
`row` seems more like a `list` than a `dict`

Reply:
it's a db row that's represented as a` dict`. I heard you though....

Reply:
>it's a db row that's represented as a dict

is it though? `row` seems to be accessed like a sequence
```py
        assert row[1] == 1, "Bad domain"
        assert row[4] == 0, "Bad class"
        return cls(dkey=row[2], fid=row[0], length=row[3])
```
iirc, fetchone() returns a tuple unless you're using a dictcursor

Reply:
I'm sorry for lying to you, Jack. You are correct. 089807a

Reply:
ok cool, I'm mostly just glad to have the validation that I'm not going crazy.

BTW, are you running a static type checker on this code? Should it have picked up on this? No action needed, I'm just curious to hear about your experiences with type checking in python.

Reply:
I'm not. I have used mypy for other stuff. It's pretty good, but you need to remember to run it as well as tests. For a personal project I have a makefile that runs pytest, mypy, and black, and fails if any of those report an error. It's pretty sweet.

Reply:
That sounds great! It would be nice to add those static checks to our CI pipelines. Perhaps something to think about with the new Cloud functions we'll be adding for Counter/LL ...